crates/matrix-sdk/src/test_utils/mocks/encryption.rs

```diff
@@ -0,0 +1,578 @@
+use std::{
+    collections::BTreeMap,
+    sync::{
+        atomic::{AtomicU32, Ordering},
+        Arc, Mutex,
+    },
+};
+
+use matrix_sdk_base::SessionMeta;
+use matrix_sdk_test::SyncResponseBuilder;
+use ruma::{
+    api::{client::keys::upload_signatures::v3::SignedKeys, MatrixVersion},
+    encryption::{CrossSigningKey, DeviceKeys, OneTimeKey},
+    owned_device_id, owned_user_id,
+    serde::Raw,
+    CrossSigningKeyId, DeviceId, OneTimeKeyAlgorithm, OwnedDeviceId, OwnedOneTimeKeyId,
+    OwnedUserId, UserId,
+};
+use serde::Serialize;
+use serde_json::json;
+use wiremock::{
+    matchers::{method, path, query_param, query_param_is_missing},
+    Mock, MockGuard, MockServer, Request, ResponseTemplate,
+};
+
+use crate::{authentication::matrix::MatrixSession, config::RequestConfig, Client, SessionTokens};
+
+#[derive(Debug, Default)]
+struct Keys {
+    device: BTreeMap<OwnedUserId, BTreeMap<String, Raw<DeviceKeys>>>,
+    master: BTreeMap<OwnedUserId, Raw<CrossSigningKey>>,
+    self_signing: BTreeMap<OwnedUserId, Raw<CrossSigningKey>>,
+    user_signing: BTreeMap<OwnedUserId, Raw<CrossSigningKey>>,
+    one_time_keys: BTreeMap<
+        OwnedUserId,
+        BTreeMap<OwnedDeviceId, BTreeMap<OwnedOneTimeKeyId, Raw<OneTimeKey>>>,
+    >,
+}
+
+/// A [`wiremock`] [`MockServer`] along with useful methods to help mocking
+/// matrix crypto API and perform integration test with encryption.
+///
+/// It implements mock endpoints for the `keys/upload`, will store the uploaded
+/// devices and serves them back for incoming `keys/query`. It is also storing
+/// and claiming one-time-keys, allowing to set up working olm sessions.
+///
+/// Adds some helpers like `exhaust_one_time_keys` that allows to simulate a
+/// client running out of otks. More can be added if needed later.
+///
+/// It works like this:
+/// * Start by creating the mock server like this [`MatrixKeysMockServer::new`].
+///   It will setup the mocks
+/// * Create your test client using [`MatrixKeysMockServer::create_client`],
+///   this is important as it will set up an access token that will allow to know
+///   what client is doing what request.
+///
+/// # Examples
+///
+/// ```
+/// # tokio_test::block_on(async {
+/// use matrix_sdk::{
+///     ruma::{device_id, user_id},
+///     test_utils::mocks::encryption::MatrixKeysMockServer,
+/// };
+/// let mock_server = MatrixKeysMockServer::new().await;
+///
+/// let (alice, bob) = mock_server.set_up_alice_and_bob_for_encryption().await;
+///
+/// let alice_bob_device = alice
+///     .encryption()
+///     .get_device(bob.user_id().unwrap(), bob.device_id().unwrap())
+///     .await
+///     .unwrap()
+///     .expect("alice sees bob's device");
+/// ```
+pub struct MatrixKeysMockServer {
+    /// The underlying [`wiremock`] [`MockServer`]
+    pub server: MockServer,
+    keys: Arc<Mutex<Keys>>,
+    token_to_user_id_map: Arc<Mutex<BTreeMap<String, OwnedUserId>>>,
+    token_counter: AtomicU32,
+}
```

Comment:
This doesn't really incorporate the `Keys` struct into `MatrixMockServer` instead it creates another mock server concept with much less functionality.

I don't think we'd want to have two of those.